### PR TITLE
Add binary filter function

### PR DESCRIPTION
**Pull request by Haaankkk**

### CODE REVIEW

**Comment by jesse-wei** [comment]
Awesome work here, Hank! Binary threshold looks good :)

Bugs that I noticed (I have to work on other hw so can't get to them rn) for commit 4c2ca6f:

1. Otsu button and binary threshold button should not both be set at the same time. We can reuse the X, Y, Z button code for this.
2. Otsu preview makes the brain black and the background white. Binary threshold preview makes the brain white and the background black. We should apply a `NotImageFilter` after Otsu in `main.render_threshold()`. I think this also means we don't need a `NotImageFilter` after binary threshold in `imgproc.contour2()`
3. For most of the binary threshold settings I've tried, I get this when clicking Apply

```text
Traceback (most recent call last):
  File "/Users/jesse/Documents/GitHub/COMP523/HeadCircumferenceTool/src/GUI/main.py", line 194, in settings_export_view_toggle
    self.render_circumference(binary_contour_slice)
  File "/Users/jesse/Documents/GitHub/COMP523/HeadCircumferenceTool/src/GUI/main.py", line 524, in render_circumference
    circumference: float = imgproc.length_of_contour(binary_contour_slice)
  File "/Users/jesse/Documents/GitHub/COMP523/HeadCircumferenceTool/src/utils/imgproc.py", line 164, in length_of_contour
    parent_contour: np.ndarray = contours[0]
IndexError: tuple index out of range
```

However, when loading the BCP image (first image) and selecting binary threshold with lower threshold 0 and upper 10, clicking apply (don't click preview) does render the contour.

Again, awesome work here! Now that the merge conflicts are resolved, it should be easier to continue work on this branch :)

**Comment by Haaankkk** [comment]
It is interesting that the second bug and the third bug actually exist. I found and fixed these two bugs before I committed. And the whole function worked well on my computer. 

The reason that the background color is flipped is that the inside value and outside value should be 0 and 1 for the binary threshold (you can find them in the global_vars). And this bug caused the third bug.

The first bug did not appear on my version. I guess it is because the UI file changed.

Thanks for the comments! No worries, let me try to locate and solve the problem first.

**Comment by jesse-wei** [comment]
I can confirm bug 1 does not exist on commit 3886574 (your first commit). Definitely one of my commits caused bug 1 because it does not exist in yours (3886574).

Not sure about the other bugs. However, I'm seeing bugs 2 and 3 on 3886574. It's strange that we're seeing different behavior on the same commit... maybe our environments are different or we're using different binary threshold settings, idk.

**Comment by Haaankkk** [comment]
global_vars.BINARY_THRESHOLD_FILTER.SetInsideValue(0)
global_vars.BINARY_THRESHOLD_FILTER.SetOutsideValue(1)

These two lines of code determine the background color. Maybe try to flip the value to see if the image is different.

**Comment by jesse-wei** [comment]
> global_vars.BINARY_THRESHOLD_FILTER.SetInsideValue(0) global_vars.BINARY_THRESHOLD_FILTER.SetOutsideValue(1)
> 
> These two lines of code determine the background color. Maybe try to flip the value to see if the image is different.

Nice, flipping these two lines in `img_helpers.initialize_globals` makes the preview look the same as the Otsu threshold (black inside, white outside) when using reasonable threshold values. This also makes contouring work.

I think we can merge after bug 1 (which came up after resolving merge conflicts) is fixed. Also, could you merge `imgproc.contour` and `imgproc.contour2` into one function? We can add a parameter `otsu_or_binary` to determine which filter to use and whether to apply a NotImageFilter.

**Comment by Haaankkk** [comment]
> > global_vars.BINARY_THRESHOLD_FILTER.SetInsideValue(0) global_vars.BINARY_THRESHOLD_FILTER.SetOutsideValue(1)
> > These two lines of code determine the background color. Maybe try to flip the value to see if the image is different.
> 
> Nice, flipping these two lines in `img_helpers.initialize_globals` makes the preview look the same as the Otsu threshold (black inside, white outside) when using reasonable threshold values. This also makes contouring work.
> 
> I think we can merge after bug 1 (which came up after resolving merge conflicts) is fixed. Also, could you merge `imgproc.contour` and `imgproc.contour2` into one function? We can add a parameter `otsu_or_binary` to determine which filter to use and whether to apply a NotImageFilter.

Sure I will fix the bug and merge the two methods together. 

It is interesting that we need the opposite values for the filter here.

File: src/utils/img_helpers.py [hunk 1]
```diff
@@ -74,6 +74,11 @@ def initialize_globals(path_list: list[Path]) -> None:
     global_vars.X_CENTER = get_middle_dimension(curr_img, View.X)
     global_vars.Y_CENTER = get_middle_dimension(curr_img, View.Y)
 
+    global_vars.BINARY_THRESHOLD_FILTER.SetLowerThreshold(100)
+    global_vars.BINARY_THRESHOLD_FILTER.SetUpperThreshold(200)
+    global_vars.BINARY_THRESHOLD_FILTER.SetInsideValue(0)
+    global_vars.BINARY_THRESHOLD_FILTER.SetOutsideValue(1)
```
Comment (jesse-wei) [inline src/utils/img_helpers.py:84]: On Hank's laptop, these two values work. On my laptop, I need to flip them like so for contouring (i.e., click Apply) to work. Very weird issue...

```py
    global_vars.BINARY_THRESHOLD_FILTER.SetInsideValue(1)
    global_vars.BINARY_THRESHOLD_FILTER.SetOutsideValue(0)
```

**Review by jesse-wei — APPROVED** [review]

File: src/utils/global_vars.py [hunk 2]
```diff
@@ -87,21 +87,6 @@
...
-"""Global Binary filter"""
-LOWER_THRESHOLD: int = 100
-"""Threshold option for binary threshold"""
-UPPER_THRESHOLD: int = 200
-"""Threshold option for binary threshold"""
-INSIDE_VALUE: int = 0
-"""Threshold option for inside color"""
-OUTSIDE_VALUE: int = 1
-"""Threshold option for outside color"""
-
```
Comment (jesse-wei) [inline src/utils/global_vars.py]: For some reason, these lines were duplicated. No idea how this got in there, maybe came from when I was resolving merge conflicts

**Comment by jesse-wei** [comment]
Remaining bugs,

1. Threshold settings should be disabled when Otsu radio button is checked, enabled when binary radio button is checked.
2. Clicking Apply (`settings_view_toggle`) does not apply the threshold settings that are currently typed in, would have to hit Preview first to apply the settings. Connect the apply button to `update_binary_filter_settings`, similar to how it's done for `update_smoothing_settings`.
3. Placeholder text for default lower threshold and upper threshold values should be displayed at program startup, not just after Preview is hit.

File: src/utils/img_helpers.py (hunk 1 above)
Comment (jesse-wei) [reply src/utils/img_helpers.py]: We were using different threshold settings. Fixed by 8eb5383, though the background color detection function could be improved.

**Comment by Haaankkk** [comment]
I updated the codes to fix these three bugs.

**Comment by jesse-wei** [comment]
Awesome work!!! Let's merge this in :)

**Comment by jesse-wei** [comment]
Deleted the branch since this seems done. Can restore if not.

**Pull request merged**